mc-crusher.c

```diff
@@ -180,305 +199,193 @@ static inline void run_counter(struct connection *c) {
+    protocol_binary_request_get *pkt = (protocol_binary_request_get *)c->wbuf_pos;
+    bin_prep_set(c);
+    pkt->message.header.request.opcode = PROTOCOL_BINARY_CMD_SETQ;
+    // Continue to send since we don't expect to read anything.
+    c->next_state = conn_sending;
 }
 
-static void write_ascii_mget_to_client(void *arg) {
+/* Unhappy with this, but it's still shorter/better than the old code.
+ * Binprot is just unwieldy in C, or I haven't figured out how to use it
+ * simply yet.
+ */
+static void bin_write_to_client(void *arg) {
     struct connection *c = arg;
-    int wbytes = 0;
-    int written = 0;
-    int i;
-    strcpy(c->wbuf, "get ");
-    written += 4;
-    for (i = 0; i < c->mget_count; i++) {
-        written += sprintf(c->wbuf + written, "%s%llu ",
-            c->key_prefix, (unsigned long long)*c->cur_key);
-        run_counter(c);
+    struct iovec *vecs = c->vecs;
+    protocol_binary_request_header *pkt = (protocol_binary_request_header *)c->wbuf_pos;
+    memset(pkt, 0, sizeof(protocol_binary_request_header));
+    pkt->request.magic = PROTOCOL_BINARY_REQ;
+    c->bin_prep_cmd(c);
+    if (c->key_prealloc) {
+        vecs[c->iov_used].iov_base = c->keys[*c->cur_key].key;
+        vecs[c->iov_used].iov_len  = c->keys[*c->cur_key].key_len;
+    } else {
+        vecs[c->iov_used].iov_base = c->wbuf_pos;
+        vecs[c->iov_used].iov_len = c->bin_format(c);
+        c->wbuf_pos += vecs[c->iov_used].iov_len;
+    }
+    int bodylen = vecs[c->iov_used].iov_len + pkt->request.extlen;
+    pkt->request.keylen = htons(vecs[c->iov_used].iov_len);
+    c->iov_used++;
+    if (c->value_size) {
+        bodylen += c->value_size;
+        vecs[c->iov_used].iov_base = c->value;
+        vecs[c->iov_used].iov_len = c->value_size;
+        c->iov_used++;
     }
-    strcpy(c->wbuf + (written), "\r\n");
-    wbytes = send(c->fd, &c->wbuf, written + 2, 0);
-    c->state = conn_reading;
+    pkt->request.bodylen = htonl(bodylen);
+    run_counter(c);
+}
+
+/* === ASCII PROTOCOL TESTS === */
+
+static int ascii_mget_format(struct connection *c) {
+    return sprintf(c->wbuf_pos, "%s%llu ", c->key_prefix,
+            (unsigned long long)*c->cur_key);
 }
 
-static void prealloc_write_ascii_mget_to_client(void *arg) {
+/* Multigets have a weird/specific format. */
+static void ascii_write_mget_to_client(void *arg) {
     struct connection *c = arg;
     int i;
     struct iovec *vecs = c->vecs;
     vecs[0].iov_base = "get ";
     vecs[0].iov_len  = 4;
     for (i = 1; i < c->mget_count + 1; i++) {
-        vecs[i].iov_base = c->keys[*c->cur_key].key;
-        vecs[i].iov_len  = c->keys[*c->cur_key].key_len;
+        if (c->key_prealloc) {
+            vecs[i].iov_base = c->keys[*c->cur_key].key;
+            vecs[i].iov_len  = c->keys[*c->cur_key].key_len;
+        } else {
+            vecs[i].iov_base = c->wbuf_pos;
+            vecs[i].iov_len  = ascii_mget_format(c);
+        }
         run_counter(c);
```

Comment:
since this function increments `cur_key` does it make sense to rename it accordingly? like `next_key`?